### PR TITLE
Expose sequence recovery data via SessionStateListener

### LINKED ISSUE #313: Expose sequence recovery data externally via the SessionStateListener interface
Today, there's no way of knowing what's the current sequence recovery state of the engine. We're are currently developing a client application that needs to react on `ResendRequest` sent to the target and `SequenceReset` messages (for hard resets) received from it.

So, my suggestion is to enhance the `SessionStateListener`  with the following methods.

```java
void onResendRequestSent(int beginSeqNo ,int endSeqNo, int currentEndSeqNo) 
```
This method will called when the send issues a ResendRequest to the target, populated by the values of the reset -equest range

```java
void onSequenceResetReceived(int newSeqNo, boolean gapFillFlag)
```

This method will be called when we recieve a SequenceReset messages, mostly here in order to allow the sender side tobe made aware of hard-resets (when the gapFillFlag is false)

```java
void onResendRequestSatisfied(int beginSeqNo ,int endSeqNo)
```

This method will be called when the quickfix engine assumes the gap has been satisfied (meaning, sequence recovery has completed)

I'd be willing to contribute a PR, should you approve this request.

### PR DESCRIPTION
**Pull request by sheinbergon**
Fixes #313

See #313 for additional details

### CODE REVIEW

**Comment by chrjohn** [comment]
Thanks for the PR!
I think this enhancement is sensible to keep the application free from logic to check for ResendRequest/SequenceReset messages in the toAdmin/fromAdmin callbacks.
We built logic to check whether our FIX process was currently waiting for a Resend to be satisfied. This could be largely simplified once this is implemented.

File: quickfixj-core/src/test/java/quickfix/SessionResetTest.java [hunk 1]
```diff
@@ -137,6 +137,16 @@ public void onMissedHeartBeat() {
...
         public void onHeartBeatTimeout() {
         }
+
+        public void onResendRequestSent(int beginSeqNo, int endSeqNo, int currentEndSeqNo) {
+        }
+
+        public void onSequenceResetReceived(int newSeqNo, boolean gapFillFlag) {
+        }
+
+        @Override
```
Comment (chrjohn) [inline quickfixj-core/src/test/java/quickfix/SessionResetTest.java:147]: nitpick: either remove `@Override` or put it to all other methods as well.

**Comment by sheinbergon** [comment]
> Thanks for the PR!
> I think this enhancement is sensible to keep the application free from logic to check for ResendRequest/SequenceReset messages in the toAdmin/fromAdmin callbacks.
> We built logic to check whether our FIX process was currently waiting for a Resend to be satisfied. This could be largely simplified once this is implemented.


Yes, actuallty that was our exact use case. I tried rolling our own logic, saw it felt and acted 'kind of iffy' and decided to solve the problem at the source instead

> 
> Edit: put to 3.0.0 milestone due to interface changes.

I'm not well aware of QFJ's release cycle/policy, but I do see that a version is released once  a year or so.
My only concern is that 3.0.0 means a release ETA of 2 years. While I have already built our own version and stored on our private nexus server, this is of course less than optimal. 

@chrjohn Is it crucial that patch versions don't contain interface changes?

File: quickfixj-core/src/test/java/quickfix/SessionResetTest.java (hunk 1 above)
Comment (sheinbergon) [reply quickfixj-core/src/test/java/quickfix/SessionResetTest.java:147]: Fixed

**Comment by chrjohn** [comment]
> > Edit: put to 3.0.0 milestone due to interface changes.
> 
> I'm not well aware of QFJ's release cycle/policy, but I do see that a version is released once a year or so.
> My only concern is that 3.0.0 means a release ETA of 2 years. While I have already built our own version and stored on our private nexus server, this is of course less than optimal.
> 
> @chrjohn Is it crucial that patch versions don't contain interface changes?

We try to follow semantic versioning. We need to increase the major version when backward incompatible changes are introduced. Since you have added new interface methods, users need to adapt (i.e. add implementations for these methods) and recompile their applications. IMHO these are backward incompatible changes. But maybe I am misinterpreting this. @the-thing or @philipwhiuk , you have any opinion on that?

You could of course add `default` implementations in the interface which would avoid recompilation. However, this looks a bit ugly in the code when `default` and non-`default` methods are mixed in the interface.

**Review by the-thing — COMMENTED** [review]
I just had a quick look. The change itself looks OK. This is an enhancement, so it probably does not require immediate release.

I'm not against releasing it with the next minor with **default** methods. They have been used in the past `quickfix.Connector`, `quickfix.MessageFactory` etc.

**Comment by chrjohn** [comment]
OK, thanks for your opinion on this @the-thing . So how about converting all methods in `SessionStateListener` to `default` methods then? People already implementing these methods shouldn't care and all others are not forced to implement the (new) methods. Plus the interface does not mix both method types.

**Comment by sheinbergon** [comment]
> OK, thanks for your opinion on this @the-thing . So how about converting all methods in `SessionStateListener` to `default` methods then? People already implementing these methods shouldn't care and all others are not forced to implement the (new) methods. Plus the interface does not mix both method types.

@chrjohn Done, all interface methods now have void defaults.

**Comment by sheinbergon** [comment]
@chrjohn Just checking. As the milestone has changed to QFJ 2.2.1, is there anything preventing us from merging this PR?

**Comment by chrjohn** [comment]
Time constraints... ;) Just wanted to take a closer look, hopefully the next days...

**Comment by chrjohn** [comment]
Just realized that @the-thing already took the time to do a review, so this is good to go.

**Pull request merged**